persistence/src/main/java/com/epam/esm/repository/TagRepositoryImpl.java

```diff
@@ -0,0 +1,89 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.entity.Tag.Columns;
+import com.epam.esm.mapper.TagMapper;
+import com.epam.esm.specification.SqlSpecification;
+import com.epam.esm.specification.TagsByCertificateIdSpecification;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.util.HashSet;
+import java.util.Set;
+
+@Repository
+public class TagRepositoryImpl extends AbstractRepository<Tag> implements TagRepository {
+    private static final String INSERT_TAG_QUERY = "INSERT INTO tags (name) VALUES (:name)";
+    private static final String ADD_TAGS_QUERY = "INSERT INTO gift_certificates_tags(gift_certificate_id, tag_id) VALUES (:gift_certificate_id, :tag_id)";
+
+    @Autowired
+    protected TagRepositoryImpl(DataSource dataSource) {
+        super(dataSource);
+    }
+
+    @Override
+    protected String getTableName() {
+        return Tag.TAGS_TABLE_NAME;
+    }
+
+    @Override
+    protected RowMapper<Tag> getRowMapper() {
+        return new TagMapper();
+    }
+
+    @Override
+    public Tag save(Tag tag) {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        MapSqlParameterSource params = new MapSqlParameterSource();
+        params.addValue(Columns.NAME.getColumn(), tag.getName());
+        template.update(INSERT_TAG_QUERY, params, keyHolder);
+        Long id = (Long) keyHolder.getKeys().get("id");
+        return getById(id).get();
+    }
+
+    @Override
+    public Set<Tag> getTagsByCertificateId(Long certificateId) {
+        SqlSpecification specification = new TagsByCertificateIdSpecification(certificateId);
+        return new HashSet<>(query(specification));
+    }
+
+    @Override
+    public Set<Tag> createNewTags(Set<Tag> tags) {
+        Set<Tag> newTags = findNewTags(tags);
+        newTags.forEach(this::save);
+        return newTags;
+    }
+
+    private Set<Tag> findNewTags(Set<Tag> tags) {
+        Set<Tag> newTags = new HashSet<>();
+        for (Tag tag : tags) {
+            if (!getByName(tag.getName()).isPresent()) {
+                newTags.add(tag);
+            }
+        }
+        return newTags;
+    }
```

Comment:
Please move code from `createNewTags` and `findNewTags` to the service layer.  The persistence layer must contain only code related to data storage interactions and not business logic.